keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1112 @@
+and make progress.
+-->
+
+This proposal does not try to address the following scenarios but relies on the
+underlying PVC resize logic to take the appropriate action or indicate error as needed.
+
+- Shrinking of volume is not allowed.
+- Few environments require Pod restarts in order to get access to the resized volume (meaning that the underlying CSI driver doesn't support online expansions). This KEP won't address such scenarios.
+- Modifying anything but the `resources.request.storage` field of the `persistentVolumeClaim` (eg, `storageClassName`, `labels`, `annotations`).
+- Resizing a PVC associated with a StatefulSet directly by modifying the PVC spec will not affect the StatefulSet `.spec.volumeClaimTemplates`.
+- Resizing ephemeral volumes.
+- Correcting failures resulting from bad input or environmental issues (such as insufficient capacity/quota). We'll make an attempt to do as much validation as we can in the API server, but cases of errors/failures past admission will have to be handled by the user. This KEP will discuss the different failures that could happen past admission.
+
+  
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Before we proceed to the proposal it is imperative to layout the various components involved
+in creating and managing the persistent volumes associated with `StatefulSet`. 
+
+- StatefulSet controller creates the PVCs from `volumeClaimTemplates` and associates them with the pods
+spawned from it. Reconciliation logic runs for every change made to the StatefulSet
+`Spec` and `Status` reflects the progress towards making the real world state equal to the
+`Spec`.
+- The validation logic in the API server is responsible for restricting the fields
+which can be edited in a StatefulSet specification.
+  - Part of the validation that the API server performs for objects (eg `StatefulSet`, `PVC`) happens in the object's API server code and part happens in an admission plugin.
+- The RBAC specification in the controller policy is responsible for adding the required 
+RBAC permission for StatefulSet.
+- The StatefulSet controller also supports revision control.
+
+The following changes are proposed to achieve the goal:
+1. Make the `volumeClaimTemplates` storage size (`.spec.resources.request.storage`) editable by modification to the api server
+   validation. We should allow the user to both increase or decrease the storage size in case of expansion failures. Allowing the user to decrease the storage size will give them the ability to address failed expansions, as described in [KEP 1790](../1790-recover-resize-failure/README.md).
```

Comment:
In non-goals section you've explicitly mentioned `Shrinking of volume is not allowed.` so I'm confused with `increase or decrease` here.

Reply:
I'll have to add more clarity here. The current plan I have in mind is to allow you to decrease the size of the PVC to the highest (across all replicas) capacity and not lower. Allowing the user to decrease the size up to the highest capacity will allow to recover from failed resizes